{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2023-05-17T04:41:05.446124Z",
     "start_time": "2023-05-17T04:41:04.263917Z"
    }
   },
   "outputs": [],
   "source": [
    "import numpy as np, pandas as pd, matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import FinancialMachineLearning as fml"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "### Chapter 8. Feature Importances\n",
    "#### Exercise 1\n",
    "8.6절의 코드를 사용하여 다음을 수행하라\n",
    "\n",
    "**(a)** Data Set$(X,y)$을 생성하라"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "outputs": [],
   "source": [
    "trnsX, const = fml.getTestData(n_features = 40, n_informative = 10, n_redundant = 10, n_samples = 10000)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-05-17T04:41:05.589447Z",
     "start_time": "2023-05-17T04:41:05.447397Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "outputs": [
    {
     "data": {
      "text/plain": "                 I_0       I_1       I_2       I_3       I_4       I_5  \\\n1985-01-17  2.843740  0.456554  0.171107 -4.511382  0.278990 -3.474726   \n1985-01-18  3.561541 -1.566097  3.342813 -1.938909  2.075749 -3.486711   \n1985-01-21  7.699248 -3.030124 -0.859302 -0.033351  1.113719 -0.877844   \n1985-01-22 -0.149801 -3.182187  2.695894  1.359997  2.992416 -0.417971   \n1985-01-23 -2.157903  0.046380  0.697217 -1.012036  1.856002 -2.311465   \n...              ...       ...       ...       ...       ...       ...   \n2023-05-11  2.489902 -0.523701  0.618210 -1.837222  1.905865  0.051481   \n2023-05-12 -0.742495  1.595855  2.173729  1.420993 -1.619853  1.952217   \n2023-05-15 -0.420406 -0.650531  3.524902 -0.608855 -0.823038  0.070312   \n2023-05-16 -0.697229 -1.974764  0.438944 -0.208492 -0.279182 -1.887733   \n2023-05-17 -3.142547 -4.576872 -1.192787  4.092471 -0.913382 -0.214720   \n\n                 I_6       I_7       I_8       I_9  ...      N_10      N_11  \\\n1985-01-17  2.955550  2.698865  1.542440  2.198168  ... -0.330515 -0.845502   \n1985-01-18  0.494908  0.309615  1.059439 -0.792433  ... -0.020384 -0.751467   \n1985-01-21  2.344033  4.089113  2.287786  0.611413  ...  0.744056  0.914181   \n1985-01-22 -1.214058  1.268313 -3.720913 -2.580578  ... -1.960632 -2.064914   \n1985-01-23  2.715493  0.444433 -1.921790 -2.472372  ... -0.841121  0.081347   \n...              ...       ...       ...       ...  ...       ...       ...   \n2023-05-11 -2.072887  1.133835  0.055833 -1.655887  ... -0.013532 -0.215027   \n2023-05-12 -2.131979 -2.340327 -1.629928  0.080561  ...  0.962720 -0.191763   \n2023-05-15 -1.619873 -1.560697 -0.939518 -1.471037  ... -0.176512  0.676007   \n2023-05-16 -1.931841 -1.163289 -0.111943 -0.732073  ...  1.179474 -0.825594   \n2023-05-17 -2.212020 -1.341850 -0.225639 -1.141243  ...  0.772330  2.041646   \n\n                N_12      N_13      N_14      N_15      N_16      N_17  \\\n1985-01-17 -1.477466  1.217536  0.304644  1.557365  0.202843  0.160110   \n1985-01-18  0.212077  0.285038  0.125461  0.203534 -0.376495 -0.938780   \n1985-01-21  1.586483  0.692802 -0.953431  0.679360  0.565153  0.219302   \n1985-01-22  1.258648 -1.031856  0.645146 -0.063900  0.305844  0.371489   \n1985-01-23 -2.587682 -0.416436 -1.077859 -0.428086 -0.183735 -0.434254   \n...              ...       ...       ...       ...       ...       ...   \n2023-05-11  0.869924 -0.932547  0.644979  0.175920  1.353286 -1.436413   \n2023-05-12  2.788231  0.055138  0.807920  0.457464  1.129258  0.273195   \n2023-05-15  0.878948  1.322274 -0.794662  0.646342  0.464202 -1.220561   \n2023-05-16 -0.419181  1.142775 -0.696596 -0.018242  0.090638  1.535604   \n2023-05-17  0.052423  1.953935 -1.333130 -1.097405  0.693821  1.509674   \n\n                N_18      N_19  \n1985-01-17  0.933805 -0.132272  \n1985-01-18 -0.142879  0.533263  \n1985-01-21 -1.110504 -1.086061  \n1985-01-22  3.218969  0.867178  \n1985-01-23 -2.124955 -0.709056  \n...              ...       ...  \n2023-05-11 -0.741075  1.054140  \n2023-05-12  1.419955 -1.095283  \n2023-05-15  0.101444  0.455718  \n2023-05-16 -0.857168  0.433795  \n2023-05-17  1.064864  1.006815  \n\n[10000 rows x 40 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>I_0</th>\n      <th>I_1</th>\n      <th>I_2</th>\n      <th>I_3</th>\n      <th>I_4</th>\n      <th>I_5</th>\n      <th>I_6</th>\n      <th>I_7</th>\n      <th>I_8</th>\n      <th>I_9</th>\n      <th>...</th>\n      <th>N_10</th>\n      <th>N_11</th>\n      <th>N_12</th>\n      <th>N_13</th>\n      <th>N_14</th>\n      <th>N_15</th>\n      <th>N_16</th>\n      <th>N_17</th>\n      <th>N_18</th>\n      <th>N_19</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1985-01-17</th>\n      <td>2.843740</td>\n      <td>0.456554</td>\n      <td>0.171107</td>\n      <td>-4.511382</td>\n      <td>0.278990</td>\n      <td>-3.474726</td>\n      <td>2.955550</td>\n      <td>2.698865</td>\n      <td>1.542440</td>\n      <td>2.198168</td>\n      <td>...</td>\n      <td>-0.330515</td>\n      <td>-0.845502</td>\n      <td>-1.477466</td>\n      <td>1.217536</td>\n      <td>0.304644</td>\n      <td>1.557365</td>\n      <td>0.202843</td>\n      <td>0.160110</td>\n      <td>0.933805</td>\n      <td>-0.132272</td>\n    </tr>\n    <tr>\n      <th>1985-01-18</th>\n      <td>3.561541</td>\n      <td>-1.566097</td>\n      <td>3.342813</td>\n      <td>-1.938909</td>\n      <td>2.075749</td>\n      <td>-3.486711</td>\n      <td>0.494908</td>\n      <td>0.309615</td>\n      <td>1.059439</td>\n      <td>-0.792433</td>\n      <td>...</td>\n      <td>-0.020384</td>\n      <td>-0.751467</td>\n      <td>0.212077</td>\n      <td>0.285038</td>\n      <td>0.125461</td>\n      <td>0.203534</td>\n      <td>-0.376495</td>\n      <td>-0.938780</td>\n      <td>-0.142879</td>\n      <td>0.533263</td>\n    </tr>\n    <tr>\n      <th>1985-01-21</th>\n      <td>7.699248</td>\n      <td>-3.030124</td>\n      <td>-0.859302</td>\n      <td>-0.033351</td>\n      <td>1.113719</td>\n      <td>-0.877844</td>\n      <td>2.344033</td>\n      <td>4.089113</td>\n      <td>2.287786</td>\n      <td>0.611413</td>\n      <td>...</td>\n      <td>0.744056</td>\n      <td>0.914181</td>\n      <td>1.586483</td>\n      <td>0.692802</td>\n      <td>-0.953431</td>\n      <td>0.679360</td>\n      <td>0.565153</td>\n      <td>0.219302</td>\n      <td>-1.110504</td>\n      <td>-1.086061</td>\n    </tr>\n    <tr>\n      <th>1985-01-22</th>\n      <td>-0.149801</td>\n      <td>-3.182187</td>\n      <td>2.695894</td>\n      <td>1.359997</td>\n      <td>2.992416</td>\n      <td>-0.417971</td>\n      <td>-1.214058</td>\n      <td>1.268313</td>\n      <td>-3.720913</td>\n      <td>-2.580578</td>\n      <td>...</td>\n      <td>-1.960632</td>\n      <td>-2.064914</td>\n      <td>1.258648</td>\n      <td>-1.031856</td>\n      <td>0.645146</td>\n      <td>-0.063900</td>\n      <td>0.305844</td>\n      <td>0.371489</td>\n      <td>3.218969</td>\n      <td>0.867178</td>\n    </tr>\n    <tr>\n      <th>1985-01-23</th>\n      <td>-2.157903</td>\n      <td>0.046380</td>\n      <td>0.697217</td>\n      <td>-1.012036</td>\n      <td>1.856002</td>\n      <td>-2.311465</td>\n      <td>2.715493</td>\n      <td>0.444433</td>\n      <td>-1.921790</td>\n      <td>-2.472372</td>\n      <td>...</td>\n      <td>-0.841121</td>\n      <td>0.081347</td>\n      <td>-2.587682</td>\n      <td>-0.416436</td>\n      <td>-1.077859</td>\n      <td>-0.428086</td>\n      <td>-0.183735</td>\n      <td>-0.434254</td>\n      <td>-2.124955</td>\n      <td>-0.709056</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>2023-05-11</th>\n      <td>2.489902</td>\n      <td>-0.523701</td>\n      <td>0.618210</td>\n      <td>-1.837222</td>\n      <td>1.905865</td>\n      <td>0.051481</td>\n      <td>-2.072887</td>\n      <td>1.133835</td>\n      <td>0.055833</td>\n      <td>-1.655887</td>\n      <td>...</td>\n      <td>-0.013532</td>\n      <td>-0.215027</td>\n      <td>0.869924</td>\n      <td>-0.932547</td>\n      <td>0.644979</td>\n      <td>0.175920</td>\n      <td>1.353286</td>\n      <td>-1.436413</td>\n      <td>-0.741075</td>\n      <td>1.054140</td>\n    </tr>\n    <tr>\n      <th>2023-05-12</th>\n      <td>-0.742495</td>\n      <td>1.595855</td>\n      <td>2.173729</td>\n      <td>1.420993</td>\n      <td>-1.619853</td>\n      <td>1.952217</td>\n      <td>-2.131979</td>\n      <td>-2.340327</td>\n      <td>-1.629928</td>\n      <td>0.080561</td>\n      <td>...</td>\n      <td>0.962720</td>\n      <td>-0.191763</td>\n      <td>2.788231</td>\n      <td>0.055138</td>\n      <td>0.807920</td>\n      <td>0.457464</td>\n      <td>1.129258</td>\n      <td>0.273195</td>\n      <td>1.419955</td>\n      <td>-1.095283</td>\n    </tr>\n    <tr>\n      <th>2023-05-15</th>\n      <td>-0.420406</td>\n      <td>-0.650531</td>\n      <td>3.524902</td>\n      <td>-0.608855</td>\n      <td>-0.823038</td>\n      <td>0.070312</td>\n      <td>-1.619873</td>\n      <td>-1.560697</td>\n      <td>-0.939518</td>\n      <td>-1.471037</td>\n      <td>...</td>\n      <td>-0.176512</td>\n      <td>0.676007</td>\n      <td>0.878948</td>\n      <td>1.322274</td>\n      <td>-0.794662</td>\n      <td>0.646342</td>\n      <td>0.464202</td>\n      <td>-1.220561</td>\n      <td>0.101444</td>\n      <td>0.455718</td>\n    </tr>\n    <tr>\n      <th>2023-05-16</th>\n      <td>-0.697229</td>\n      <td>-1.974764</td>\n      <td>0.438944</td>\n      <td>-0.208492</td>\n      <td>-0.279182</td>\n      <td>-1.887733</td>\n      <td>-1.931841</td>\n      <td>-1.163289</td>\n      <td>-0.111943</td>\n      <td>-0.732073</td>\n      <td>...</td>\n      <td>1.179474</td>\n      <td>-0.825594</td>\n      <td>-0.419181</td>\n      <td>1.142775</td>\n      <td>-0.696596</td>\n      <td>-0.018242</td>\n      <td>0.090638</td>\n      <td>1.535604</td>\n      <td>-0.857168</td>\n      <td>0.433795</td>\n    </tr>\n    <tr>\n      <th>2023-05-17</th>\n      <td>-3.142547</td>\n      <td>-4.576872</td>\n      <td>-1.192787</td>\n      <td>4.092471</td>\n      <td>-0.913382</td>\n      <td>-0.214720</td>\n      <td>-2.212020</td>\n      <td>-1.341850</td>\n      <td>-0.225639</td>\n      <td>-1.141243</td>\n      <td>...</td>\n      <td>0.772330</td>\n      <td>2.041646</td>\n      <td>0.052423</td>\n      <td>1.953935</td>\n      <td>-1.333130</td>\n      <td>-1.097405</td>\n      <td>0.693821</td>\n      <td>1.509674</td>\n      <td>1.064864</td>\n      <td>1.006815</td>\n    </tr>\n  </tbody>\n</table>\n<p>10000 rows × 40 columns</p>\n</div>"
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trnsX"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-05-17T04:41:05.616491Z",
     "start_time": "2023-05-17T04:41:05.595979Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "outputs": [
    {
     "data": {
      "text/plain": "            bin       w         t1\n1985-01-17    0  0.0001 1985-01-17\n1985-01-18    0  0.0001 1985-01-18\n1985-01-21    0  0.0001 1985-01-21\n1985-01-22    0  0.0001 1985-01-22\n1985-01-23    0  0.0001 1985-01-23\n...         ...     ...        ...\n2023-05-11    1  0.0001 2023-05-11\n2023-05-12    1  0.0001 2023-05-12\n2023-05-15    1  0.0001 2023-05-15\n2023-05-16    1  0.0001 2023-05-16\n2023-05-17    1  0.0001 2023-05-17\n\n[10000 rows x 3 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>bin</th>\n      <th>w</th>\n      <th>t1</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>1985-01-17</th>\n      <td>0</td>\n      <td>0.0001</td>\n      <td>1985-01-17</td>\n    </tr>\n    <tr>\n      <th>1985-01-18</th>\n      <td>0</td>\n      <td>0.0001</td>\n      <td>1985-01-18</td>\n    </tr>\n    <tr>\n      <th>1985-01-21</th>\n      <td>0</td>\n      <td>0.0001</td>\n      <td>1985-01-21</td>\n    </tr>\n    <tr>\n      <th>1985-01-22</th>\n      <td>0</td>\n      <td>0.0001</td>\n      <td>1985-01-22</td>\n    </tr>\n    <tr>\n      <th>1985-01-23</th>\n      <td>0</td>\n      <td>0.0001</td>\n      <td>1985-01-23</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>2023-05-11</th>\n      <td>1</td>\n      <td>0.0001</td>\n      <td>2023-05-11</td>\n    </tr>\n    <tr>\n      <th>2023-05-12</th>\n      <td>1</td>\n      <td>0.0001</td>\n      <td>2023-05-12</td>\n    </tr>\n    <tr>\n      <th>2023-05-15</th>\n      <td>1</td>\n      <td>0.0001</td>\n      <td>2023-05-15</td>\n    </tr>\n    <tr>\n      <th>2023-05-16</th>\n      <td>1</td>\n      <td>0.0001</td>\n      <td>2023-05-16</td>\n    </tr>\n    <tr>\n      <th>2023-05-17</th>\n      <td>1</td>\n      <td>0.0001</td>\n      <td>2023-05-17</td>\n    </tr>\n  </tbody>\n</table>\n<p>10000 rows × 3 columns</p>\n</div>"
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "const"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-05-17T04:41:05.622998Z",
     "start_time": "2023-05-17T04:41:05.619914Z"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process SpawnPoolWorker-26:\n",
      "Process SpawnPoolWorker-25:\n",
      "Process SpawnPoolWorker-27:\n",
      "Process SpawnPoolWorker-23:\n",
      "Process SpawnPoolWorker-17:\n",
      "Process SpawnPoolWorker-24:\n",
      "Process SpawnPoolWorker-29:\n",
      "Process SpawnPoolWorker-20:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 32, in <module>\n",
      "    from numba import jit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numba/__init__.py\", line 19, in <module>\n",
      "    from numba.core import config\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numba/core/config.py\", line 15, in <module>\n",
      "    import llvmlite.binding as ll\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/llvmlite/binding/__init__.py\", line 4, in <module>\n",
      "    from .dylib import *\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/llvmlite/binding/dylib.py\", line 3, in <module>\n",
      "    from llvmlite.binding import ffi\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/llvmlite/binding/ffi.py\", line 175, in <module>\n",
      "    from pkg_resources import resource_filename\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/__init__.py\", line 82, in <module>\n",
      "    from pkg_resources.extern.jaraco.text import (\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/_vendor/jaraco/text/__init__.py\", line 11, in <module>\n",
      "    from pkg_resources.extern.jaraco.functools import compose, method_cache\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/_vendor/jaraco/functools.py\", line 8, in <module>\n",
      "    import pkg_resources.extern.more_itertools\n",
      "  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 986, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 666, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 565, in module_from_spec\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/extern/__init__.py\", line 52, in create_module\n",
      "    return self.load_module(spec.name)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/extern/__init__.py\", line 37, in load_module\n",
      "    __import__(extant)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 986, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 680, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 846, in exec_module\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 941, in get_code\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 1040, in get_data\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 32, in <module>\n",
      "    from numba import jit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numba/__init__.py\", line 19, in <module>\n",
      "    from numba.core import config\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numba/core/config.py\", line 15, in <module>\n",
      "    import llvmlite.binding as ll\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/llvmlite/binding/__init__.py\", line 4, in <module>\n",
      "    from .dylib import *\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/llvmlite/binding/dylib.py\", line 3, in <module>\n",
      "    from llvmlite.binding import ffi\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/llvmlite/binding/ffi.py\", line 175, in <module>\n",
      "    from pkg_resources import resource_filename\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/__init__.py\", line 82, in <module>\n",
      "    from pkg_resources.extern.jaraco.text import (\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/_vendor/jaraco/text/__init__.py\", line 11, in <module>\n",
      "    from pkg_resources.extern.jaraco.functools import compose, method_cache\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/_vendor/jaraco/functools.py\", line 8, in <module>\n",
      "    import pkg_resources.extern.more_itertools\n",
      "  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 986, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 666, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 565, in module_from_spec\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/extern/__init__.py\", line 52, in create_module\n",
      "    return self.load_module(spec.name)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/extern/__init__.py\", line 37, in load_module\n",
      "    __import__(extant)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pkg_resources/_vendor/more_itertools/__init__.py\", line 3, in <module>\n",
      "    from .more import *  # noqa\n",
      "  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 982, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 925, in _find_spec\n",
      "Process SpawnPoolWorker-21:\n",
      "KeyboardInterrupt\n",
      "Process SpawnPoolWorker-22:\n",
      "Process SpawnPoolWorker-15:\n",
      "Process SpawnPoolWorker-19:\n",
      "Process SpawnPoolWorker-18:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 8, in <module>\n",
      "    import seaborn as sns\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/seaborn/__init__.py\", line 2, in <module>\n",
      "    from .rcmod import *  # noqa: F401,F403\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/seaborn/rcmod.py\", line 7, in <module>\n",
      "    from . import palettes\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/seaborn/palettes.py\", line 9, in <module>\n",
      "    from .utils import desaturate, get_color_cycle\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/seaborn/utils.py\", line 10, in <module>\n",
      "    from scipy import stats\n",
      "  File \"<frozen importlib._bootstrap>\", line 1055, in _handle_fromlist\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/scipy/__init__.py\", line 200, in __getattr__\n",
      "    return _importlib.import_module(f'scipy.{name}')\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/__init__.py\", line 485, in <module>\n",
      "Exception ignored in: <module 'threading' from '/Users/tommylee/opt/anaconda3/lib/python3.9/threading.py'>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/threading.py\", line 1428, in _shutdown\n",
      "    from ._stats_py import *\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/_stats_py.py\", line 46, in <module>\n",
      "Exception ignored in: <module 'threading' from '/Users/tommylee/opt/anaconda3/lib/python3.9/threading.py'>\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/threading.py\", line 1428, in _shutdown\n",
      "    from . import distributions\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/distributions.py\", line 8, in <module>\n",
      "    from ._distn_infrastructure import (rv_discrete, rv_continuous, rv_frozen)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/_distn_infrastructure.py\", line 25, in <module>\n",
      "    from scipy import integrate\n",
      "  File \"<frozen importlib._bootstrap>\", line 1055, in _handle_fromlist\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/scipy/__init__.py\", line 200, in __getattr__\n",
      "    return _importlib.import_module(f'scipy.{name}')\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/__init__.py\", line 101, in <module>\n",
      "    from . import dop, lsoda, vode, odepack, quadpack\n",
      "  File \"<frozen importlib._bootstrap>\", line 1007, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 986, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 680, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 846, in exec_module\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 950, in get_code\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 585, in _classify_pyc\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 81, in _unpack_uint32\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "Process SpawnPoolWorker-16:\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    def _shutdown():\n",
      "KeyboardInterrupt: \n",
      "    def _shutdown():\n",
      "KeyboardInterrupt: \n",
      "Traceback (most recent call last):\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 366, in extract\n",
      "    f.line\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 288, in line\n",
      "    self._line = linecache.getline(self.filename, self.lineno).strip()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 30, in getline\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 365, in get\n",
      "    res = self._reader.recv_bytes()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/connection.py\", line 221, in recv_bytes\n",
      "    lines = getlines(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 46, in getlines\n",
      "    buf = self._recv_bytes(maxlength)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/connection.py\", line 419, in _recv_bytes\n",
      "    return updatecache(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 136, in updatecache\n",
      "    with tokenize.open(fullname) as fp:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 394, in open\n",
      "    buf = self._recv(4)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/connection.py\", line 384, in _recv\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    chunk = read(handle, remaining)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    encoding, lines = detect_encoding(buffer.readline)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 363, in detect_encoding\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    first = read_or_stop()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 321, in read_or_stop\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    return readline()\n",
      "KeyboardInterrupt\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 366, in extract\n",
      "    f.line\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 288, in line\n",
      "    self._line = linecache.getline(self.filename, self.lineno).strip()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 30, in getline\n",
      "    lines = getlines(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 46, in getlines\n",
      "    return updatecache(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 136, in updatecache\n",
      "    with tokenize.open(fullname) as fp:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 394, in open\n",
      "    encoding, lines = detect_encoding(buffer.readline)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 313, in detect_encoding\n",
      "    filename = readline.__self__.name\n",
      "KeyboardInterrupt\n",
      "Process SpawnPoolWorker-31:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "Process SpawnPoolWorker-30:\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "Process SpawnPoolWorker-32:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.\n",
      "KeyboardInterrupt\n",
      "py\", line 2014, in expandCall\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    result = ImmediateResult(func)\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    self.results = batch()\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    return self.function(*args, **kwargs)\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/mu\n",
      "ltiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 364, in get\n",
      "    with self._rlock:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 364, in get\n",
      "    with self._rlock:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "KeyboardInterrupt\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 220, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 220, in fit\n",
      "    expanded_class_weight = compute_sample_weight(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/class_weight.py\", line 167, in compute_sample_weight\n",
      "    super().fit(\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "    expanded_class_weight = compute_sample_weight(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/class_weight.py\", line 142, in compute_sample_weight\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"<string>\", line 1, in <module>\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "    classes_full = np.unique(y_full)\n",
      "  File \"<__array_function__ internals>\", line 5, in unique\n",
      "    weight_k = compute_class_weight(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/class_weight.py\", line 48, in compute_class_weight\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    y_ind = le.fit_transform(y)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/preprocessing/_label.py\", line 116, in fit_transform\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 366, in extract\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    f.line\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 288, in line\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    self._line = linecache.getline(self.filename, self.lineno).strip()\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 350, in extract\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 30, in getline\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    lines = getlines(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 46, in getlines\n",
      "    name = co.co_name\n",
      "KeyboardInterrupt\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 366, in extract\n",
      "    return updatecache(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 93, in updatecache\n",
      "    f.line\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 288, in line\n",
      "    stat = os.stat(fullname)\n",
      "KeyboardInterrupt\n",
      "    self._line = linecache.getline(self.filename, self.lineno).strip()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 30, in getline\n",
      "    lines = getlines(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 46, in getlines\n",
      "    return updatecache(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 137, in updatecache\n",
      "    lines = fp.readlines()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/codecs.py\", line 319, in decode\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/arraysetops.py\", line 272, in unique\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 366, in extract\n",
      "    f.line\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 288, in line\n",
      "    self._line = linecache.getline(self.filename, self.lineno).strip()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 30, in getline\n",
      "    lines = getlines(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 46, in getlines\n",
      "    return updatecache(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 137, in updatecache\n",
      "    lines = fp.readlines()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/codecs.py\", line 319, in decode\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 330, in _bootstrap\n",
      "    sys.stderr.write('Process %s:\\n' % self.name)\n",
      "KeyboardInterrupt\n",
      "    ret = _unique1d(ar, return_index, return_inverse, return_counts)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/arraysetops.py\", line 333, in _unique1d\n",
      "    ar.sort()\n",
      "    def decode(self, input, final=False):\n",
      "KeyboardInterrupt\n",
      "    def decode(self, input, final=False):\n",
      "KeyboardInterrupt\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 364, in get\n",
      "    with self._rlock:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "Traceback (most recent call last):\n",
      "    self.classes_, y = _unique(y, return_inverse=True)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "KeyboardInterrupt\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/_encode.py\", line 33, in _unique\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 318, in _bootstrap\n",
      "    util._exit_function()\n",
      "KeyboardInterrupt\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 103, in print_exception\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 366, in extract\n",
      "    for line in TracebackException(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 517, in __init__\n",
      "    f.line\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 288, in line\n",
      "    self.stack = StackSummary.extract(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 353, in extract\n",
      "    self._line = linecache.getline(self.filename, self.lineno).strip()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 30, in getline\n",
      "    lines = getlines(filename, module_globals)\n",
      "    linecache.lazycache(filename, f.f_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 46, in getlines\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 171, in lazycache\n",
      "    return updatecache(filename, module_globals)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/linecache.py\", line 136, in updatecache\n",
      "    get_source = getattr(loader, 'get_source', None)\n",
      "KeyboardInterrupt\n",
      "    with tokenize.open(fullname) as fp:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 394, in open\n",
      "    encoding, lines = detect_encoding(buffer.readline)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 371, in detect_encoding\n",
      "    encoding = find_cookie(first)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/tokenize.py\", line 326, in find_cookie\n",
      "    try:\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 318, in _bootstrap\n",
      "    util._exit_function()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/util.py\", line 350, in _exit_function\n",
      "    for p in active_children():\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 47, in active_children\n",
      "    _cleanup()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 63, in _cleanup\n",
      "    for p in list(_children):\n",
      "KeyboardInterrupt\n",
      "    out = np.unique(values, return_inverse=return_inverse)\n",
      "  File \"<__array_function__ internals>\", line 5, in unique\n",
      "Process SpawnPoolWorker-13:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/arraysetops.py\", line 272, in unique\n",
      "    ret = _unique1d(ar, return_index, return_inverse, return_counts)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/arraysetops.py\", line 356, in _unique1d\n",
      "    inv_idx[perm] = imask\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    traceback.print_exc()\n",
      "KeyboardInterrupt\n",
      "Process SpawnPoolWorker-12:\n",
      "Process LokyProcess-4:\n",
      "Process LokyProcess-2:\n",
      "Process SpawnPoolWorker-14:\n",
      "Process LokyProcess-7:\n",
      "Process LokyProcess-8:\n",
      "Process LokyProcess-6:\n",
      "Process LokyProcess-5:\n",
      "Process LokyProcess-1:\n",
      "Process SpawnPoolWorker-10:\n",
      "Process SpawnPoolWorker-9:\n",
      "Process SpawnPoolWorker-11:\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "Process SpawnPoolWorker-28:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/_config.py\", line 177, in config_context\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 114, in worker\n",
      "    task = get()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/queues.py\", line 367, in get\n",
      "    return _ForkingPickler.loads(res)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 5, in <module>\n",
      "    import pandas as pd\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/pandas/__init__.py\", line 11, in <module>\n",
      "    __import__(dependency)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numpy/__init__.py\", line 337, in <module>\n",
      "    __mkl_version__ = \"{MajorVersion}.{MinorVersion}.{UpdateVersion}\".format(**mkl.get_version())\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    return [func(*args, **kwargs)\n",
      "    return self.function(*args, **kwargs)\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    return [func(*args, **kwargs)\n",
      "    yield\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return self.function(*args, **kwargs)\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "    return [func(*args, **kwargs)\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    super().fit(\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    super().fit(\n",
      "    self.results = batch()\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 214, in fit\n",
      "    super().fit(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 165, in fit\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    X, y = self._validate_data(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/base.py\", line 578, in _validate_data\n",
      "KeyboardInterrupt\n",
      "    classes_k, y_encoded[:, k] = np.unique(y[:, k], return_inverse=True)\n",
      "    X = check_array(X, **check_X_params)\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<__array_function__ internals>\", line 5, in unique\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/validation.py\", line 734, in check_array\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 123, in _parallel_build_estimators\n",
      "    with warnings.catch_warnings():\n",
      "    self.run()\n",
      "    estimator.fit(X[:, features], y, sample_weight=curr_sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/warnings.py\", line 459, in __enter__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/arraysetops.py\", line 272, in unique\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 937, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "    ret = _unique1d(ar, return_index, return_inverse, return_counts)\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "    super().fit(\n",
      "    if self._entered:\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/tree/_classes.py\", line 420, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/numpy/lib/arraysetops.py\", line 356, in _unique1d\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    result = (True, func(*args, **kwds))\n",
      "    inv_idx[perm] = imask\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    builder.build(self.tree_, X, y, sample_weight)\n",
      "    result = (True, func(*args, **kwds))\n",
      "    result = (True, func(*args, **kwds))\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "    self.run()\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "KeyboardInterrupt\n",
      "    result = (True, func(*args, **kwds))\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "    self.run()\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    result = (True, func(*args, **kwds))\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 2014, in expandCall\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    out = func(**kargs)  # put the arguments into the function\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1488, in auxFeatImpSFI\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    self._dispatch(tasks)\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    result = ImmediateResult(func)\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    self._dispatch(tasks)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    self.results = batch()\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "    return [func(*args, **kwargs)\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    self._dispatch(tasks)\n",
      "    df0 = cvScore(clf, X=trnsX[[featName]], y=cont['bin'], sample_weight=cont['w'], scoring=scoring, cvGen=cvGen)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "  File \"/Users/tommylee/Desktop/Pycharm/Financial_Machine_Learning/FinancialMachineLearning.py\", line 1310, in cvScore\n",
      "    return [func(*args, **kwargs)\n",
      "    return [func(*args, **kwargs)\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    fit = clf.fit(X = pd.DataFrame(X).iloc[train, :], y = pd.DataFrame(y).iloc[train],\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 269, in fit\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "    return self.function(*args, **kwargs)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "    return self.function(*args, **kwargs)\n",
      "    self._dispatch(tasks)\n",
      "    return self._fit(X, y, self.max_samples, sample_weight=sample_weight)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 779, in _dispatch\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/ensemble/_bagging.py\", line 394, in _fit\n",
      "    result = ImmediateResult(func)\n",
      "    result = ImmediateResult(func)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/contextlib.py\", line 137, in __exit__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "    all_results = Parallel(\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 1043, in __call__\n",
      "    self.gen.throw(typ, value, traceback)\n",
      "    job = self._backend.apply_async(batch, callback=cb)\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 208, in apply_async\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "    result = ImmediateResult(func)\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/_parallel_backends.py\", line 572, in __init__\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    if self.dispatch_one_batch(iterator):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 861, in dispatch_one_batch\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in <listcomp>\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    self.results = batch()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 262, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/utils/fixes.py\", line 216, in __call__\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 219, in __exit__\n",
      "    self._dispatch(tasks)\n",
      "    return [func(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 784, in _dispatch\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/joblib/parallel.py\", line 220, in __exit__\n",
      "    return self.function(*args, **kwargs)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/contextlib.py\", line 137, in __exit__\n",
      "    def __exit__(self, type, value, traceback):\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"<string>\", line 1, in <module>\n",
      "    self.unregister()\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "    self.gen.throw(typ, value, traceback)\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/_config.py\", line 179, in config_context\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    self._jobs.insert(job_idx, job)\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "    set_config(**old_config)\n",
      "  File \"<string>\", line 1, in <module>\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/site-packages/sklearn/_config.py\", line 88, in set_config\n",
      "    local_config = _get_threadlocal_config()\n",
      "KeyboardInterrupt\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "  File \"<string>\", line 1, in <module>\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 105, in print_exception\n",
      "    print(line, file=file, end=\"\")\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 105, in print_exception\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 105, in print_exception\n",
      "    print(line, file=file, end=\"\")\n",
      "KeyboardInterrupt\n",
      "    print(line, file=file, end=\"\")\n",
      "KeyboardInterrupt\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 116, in spawn_main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    exitcode = _main(fd, parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/spawn.py\", line 129, in _main\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "    return self._bootstrap(parent_sentinel)\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/multiprocessing/process.py\", line 331, in _bootstrap\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 105, in print_exception\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    traceback.print_exc()\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 163, in print_exc\n",
      "    print(line, file=file, end=\"\")\n",
      "KeyboardInterrupt\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 105, in print_exception\n",
      "    print_exception(*sys.exc_info(), limit=limit, file=file, chain=chain)\n",
      "  File \"/Users/tommylee/opt/anaconda3/lib/python3.9/traceback.py\", line 105, in print_exception\n",
      "    print(line, file=file, end=\"\")\n",
      "KeyboardInterrupt\n",
      "    print(line, file=file, end=\"\")\n",
      "KeyboardInterrupt\n"
     ]
    }
   ],
   "source": [
    "imp, oob, oos = fml.featImportance(trnsX, const)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-05-17T04:42:34.779419Z",
     "start_time": "2023-05-17T04:41:05.623265Z"
    }
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "**(b)** PCA변환을 $X$에 적용하라. 이를 $\\dot{X}$라 표기한다."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
